spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java

```diff
@@ -0,0 +1,502 @@
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+
+    List<List<SimpleRecord>> expected = getTestDataForMultipleSnapshots();
+    appendData(expected, location);
+
+    table.refresh();
+
+    try {
+      Dataset<Row> df = spark.readStream()
+              .format("iceberg")
+              .load(location.toString());
+      List<SimpleRecord> actual = processStreamTillEnd(df);
+
+      Assert.assertEquals(
+          expected.stream().flatMap(List::stream).collect(Collectors.toList()),
+          actual);
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testResumingStreamReadFromCheckpoint() throws IOException, TimeoutException, StreamingQueryException {
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+    File writerCheckpoint = new File(parent, "writer-checkpoint");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+    final String tempView = "microBatchView";
+
+    try {
+      Dataset<Row> df = spark.readStream()
+          .format("iceberg")
+          .load(location.toString());
+      DataStreamWriter<Row> singleBatchWriter = df.writeStream()
+          .trigger(Trigger.Once())
+          .option("checkpointLocation", writerCheckpoint.toString())
+          .foreachBatch((batchDF, batchId) -> {
+            batchDF.createOrReplaceGlobalTempView(tempView);
+          });
+
+      String globalTempView = "global_temp." + tempView;
+
+      List<SimpleRecord> processStreamOnEmptyIcebergTable = processMicroBatch(singleBatchWriter, globalTempView);
+      Assert.assertEquals(
+          Collections.emptyList(),
+          processStreamOnEmptyIcebergTable);
+
+      for (List<List<SimpleRecord>> expectedCheckpoint : getTestDataForMultipleWritesWithMultipleSnapshots()) {
+        appendData(expectedCheckpoint, location);
+        table.refresh();
+
+        List<SimpleRecord> actualDataInCurrentMicroBatch = processMicroBatch(singleBatchWriter, globalTempView);
+        Assert.assertEquals(
+            expectedCheckpoint.stream().flatMap(List::stream).collect(Collectors.toList()),
+            actualDataInCurrentMicroBatch);
+      }
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testParquetOrcAvroDataInOneTable() throws IOException, TimeoutException {
```

Comment:
It's okay to have this, but I don't see much of a benefit to testing this. File formats are orthogonal to the streaming that is done in this PR.

Reply:
a single table - with multiple file formats is a very critical scenario - for our usecase & hence the coverage. 
I found an issue with my initial implementation with this test case.